### PR TITLE
Adds support for dynamic template data in personalizations

### LINKED ISSUE #591: Add Dynamic Template Support
#### Issue Summary

On 7/24/2018, our team publicly launched dynamic content for transactional templates. It is now available for all customers sending over v3 of our Mail Send API. Iterate over lists, handle conditionals and more, thanks to native support for a subset of Handlebars syntax!

More information can be found in our [blog post announcement](https://sendgrid.com/blog/how-to-use-sendgrids-dynamic-templates-for-your-transactional-emails/).

You can currently use this feature by manually creating the request body as [shown here](https://github.com/sendgrid/sendgrid-python/blob/master/USAGE.md#v3-mail-send). 

Now, we need to create helper code and examples for this SDK.

#### Acceptance Criteria

* [Implement a helper similar to what we have for the legacy templates](https://github.com/sendgrid/sendgrid-python/blob/master/use_cases/transational_templates.md)
* Update the [transactional_templates.md](https://github.com/sendgrid/sendgrid-python/blob/master/use_cases/transational_templates.md) example to demonstrate the new *Dynamic Templates* using the helper and re-name the current example to *Legacy*

#### Documentation
- [Reference Implementation](https://github.com/sendgrid/sendgrid-nodejs/pull/691#issuecomment-407490342)
- [How to Send an Email With Dynamic Transactional Templates](https://sendgrid.com/docs/User_Guide/Transactional_Templates/how_to_send_an_email_with_transactional_templates.html)
- [Using Handlebars](https://sendgrid.com/docs/User_Guide/Transactional_Templates/Using_handlebars.html)
- [Interactive API Docs](https://dynamic-templates.api-docs.io/3.0/mail-send-with-dynamic-transactional-templates/v3-mail-send)

### PR DESCRIPTION
**Pull request by 3lnc**
### Checklist
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added in line documentation to the code I modified

### Short description of what this PR does:
Adds support for dynamic templates

Fixes #591

### CODE REVIEW

**Comment by af4ro** [comment]
Thank you so much for your PR @3lnc! 
This is a great PR and it covers all the aspects required by the specifications.

File: use_cases/transational_templates.md [hunk 1]
```diff
@@ -66,6 +66,38 @@ print(response.body)
 print(response.headers)
 ```
 
+### With dynamic templates
```
Comment (af4ro) [inline use_cases/transational_templates.md]: Might be missing dynamic content updates in the rest of the templates file

Comment (af4ro) [reply use_cases/transational_templates.md]: You can check out some of the changes made [here](https://github.com/sendgrid/sendgrid-python/pull/597/files#diff-fa3865ac405555d68b7dd54aff4ea7ff) in #597

Comment (3lnc) [reply use_cases/transational_templates.md]: Regarding updates for rest of file: this is left on purpose, as old template system is still in place, as well as code to support it. So, I guess we still need documentation for it.

Or I'm missing the point

**Comment by thinkingserious** [comment]
Hello @3lnc,<br /><br/>Thanks again for the PR!<br />                                              <br />We appreciate your contribution and look forward to                                              continued collaboration. Thanks!                                             <br /><br />Team SendGrid DX

File: sendgrid/helpers/mail/personalization.py [hunk 2]
```diff
@@ -1,6 +1,10 @@
 class Personalization(object):
     """A Personalization defines who should receive an individual message and
     how that message should be handled.
+
+    :var dynamic_template_data: data for dynamic transactional template.
+        Should be JSON-serializeable structure. No pre-processing sill be done
```
Comment (pjdelport) [inline sendgrid/helpers/mail/personalization.py]: Typo: _sill_ → _will_

Comment (thinkingserious) [reply sendgrid/helpers/mail/personalization.py]: Thanks! Fixed :)

**Comment by djw27** [comment]
Should this not also remove the use of `Substitution` from `mail_example.py` since this has been deprecated?

I also think this is a different type of interface to existing personalizations etc so could be confusing to newcomers as to why `dynamic_template_data` is added differently, i.e. setting the attribute directly on the personalization rather than using a setter like all other attributes? 

Just my thoughts...

**Comment by 3lnc** [comment]
@djw27 by the time this PR was opened I had no info on deprecation of substitution. Could you please provide reference for this info?

Regarding descriptors interface ­— personally, I really dislike the current state of codebase being overcrowded with getters. This is really not pythonic way, as there's no setters with custom logic. Hiding attributes with `_name` + `@property` is actually gives nothing but pain — in case you need to modify helper objects dynamically. But that's IMO only.

Regarding "raw" implementation of `dynamic_template_data` — this is conscious decision. Dynamic templates requires json. Which is pretty "technical". I mean, if json is needed for handlebars, you probably know what json is and how to map json into python datastructures. All the validation we can do on python side is done while serializing.

Making this interface via setters will require class(es) with recurrent references — all this just to make nested structure. Taking into account that data classes is almost wrappers around dictionaries — I'm pretty sure that raw implementation is better way.

**Comment by thinkingserious** [comment]
Thanks for the feedback @djw27 and @3lnc!

I'm currently working on an update to this SDK where your feedback would be greatly helpful. If you have time, please [take a look](https://github.com/sendgrid/sendgrid-python/issues/347).

With Best Regards,

Elmer

**Comment by djw27** [comment]
@3lnc I 100% agree with your points! I don't blame you not knowing that substitutions are being deprecated, I stumbled across the documentation by chance for dynamic templates: https://dynamic-templates.api-docs.io/3.0/mail-send-with-dynamic-transactional-templates/v3-mail-send

I guess it's a wider question for @thinkingserious, but at stage it would be pretty difficult to completely overhaul the interface used for this library

**Comment by Corey-Evans** [comment]
@thinkingserious Are substitutions still being deprecated? I remember reading this a few months back. However, now I can't find any reference to this (including on the link referenced above).

**Pull request merged**